blog/2020-10-01-hs-to-coq/2020-10-01-hs-to-coq.md

```diff
@@ -0,0 +1,1084 @@
+type signatures:
+
+```haskell
+empty :: Queue a
+push :: a -> Queue a -> Queue a
+pop :: Queue a -> Maybe (a, Queue a)
+```
+
+There are many ways to implement a queue. Here, we will use a classical
+implementation using two lists.
+
+The main idea is to maintain two parts of the queue separately:
+the *front* and the *back*.
+
+<div class="tuto">
+```haskell
+data Queue a = Queue [a]   [a]
+            {- Queue front back -}
+
+empty :: Queue a
+empty = Queue [] []
+```
+</div>
+
+We *push* an element `x` with a simple *cons* `(:)` to the back.
+
+<div class="tuto">
+```haskell
+push :: a -> Queue a -> Queue a
+push x (Queue front back) = Queue front (x : back)
+```
+</div>
+
+Elements can be *popped* by pattern-matching on the front. We remove
+the head `y` from the front if it is not empty. If the front is empty, we
+reverse the back, and it becomes the new front, from which we remove the first
+element. We fail with `Nothing` if both the front and back are empty.
+
+<div class="tuto">
+```haskell
+pop :: Queue a -> Maybe (a, Queue a)
+pop (Queue (y : front) back) = Just (y, Queue front back)
+pop (Queue [] back)
+  = case reverse back of
+      [] -> Nothing
+      y : front -> Just (y, Queue front [])
+```
+</div>
+
+This implementation ensures that the cost of `pop` is *amortized*
+across many operations, so the total cost per operation is constant on average.
+
+A detailed analysis of this structure can be found in Chris Okasaki's thesis
+([*Purely Functional Data Structures*](https://www.cs.cmu.edu/~rwh/theses/okasaki.pdf) (1996),
+Section 3.1.1),
+which includes many more examples of purely functional data structures.
+
+For more convenient testing, let's define an auxiliary function to
+remove all elements from a queue, until it comes out empty.
+
+<div class="tuto">
+```haskell
+toList :: Queue a -> [a]
+toList q
+  = case pop q of
+      Nothing -> []
+      Just (y, q') -> y : toList q'
+```
+</div>
+
+Put all of that Haskell code in a file `src/Queue.hs`.
+We can give it a quick try in ghci:
+
+<div class="tuto">
+```
+ghci src/Queue.hs                             -- Start ghci
+ghci> let q = push 3 (push 2 (push 1 empty))  -- Use push
+ghci> toList q                                -- Use pop until the end
+[1,2,3]                                       -- Output
+```
+</div>
+
+<div class="tuto">
+<details>
+<summary>Complete Haskell module `Queue.hs`</summary>
+```haskell
+module Queue where
+
+data Queue a = Queue [a]   [a]
+            {- Queue front back -}
+
+empty :: Queue a
+empty = Queue [] []
+
+push :: a -> Queue a -> Queue a
+push x (Queue front back) = Queue front (x : back)
+
+pop :: Queue a -> Maybe (a, Queue a)
+pop (Queue (y : front) back) = Just (y, Queue front back)
+pop (Queue [] back)
+  = case reverse back of
+      [] -> Nothing
+      y : front -> Just (y, Queue front [])
+
+toList :: Queue a -> [a]
+toList q
+  = case pop q of
+      Nothing -> []
+      Just (y, q') -> y : toList q'
```

Comment:
Any particular reason that you define `toList` in this way? For example, you can also define it using some way similar to `models` relation you discuss later:
```haskell
toList (Queue front back) = front ++ reverse back
```

Reply:
Also, if you are defining this function for demonstration purpose, you can also use `deriving Show` or define your own `Show` instance in the Haskell file. Because you are using the edits file from `base`, `Show` will not be translated.

Reply:
The point of `toList` is as a shorthand for using `pop` several times, because the pattern-matching is a little cumbersome to do by hand.